File: src/leetcode/P149_MaxPointsOnALine.java
package leetcode;

import java.util.HashMap;

/**
 * 	Given n points on a 2D plane, find the maximum number of 
 * 	points that lie on the same straight line.
 */

/**
 * @author      zxwtry
 * @email       zxwtry@qq.com
 * @project     OJ
 * @package     leetcode
 * @file        P149_MaxPointsOnALine.java
 * @type        P149_MaxPointsOnALine
 * @date        2016年12月30日 上午10:26:23
 * @details     Solution:  AC  36ms 45.82%
 * @details     Solution2: AC  47ms 28.50%
 */
public class P149_MaxPointsOnALine {
	static class Solution {
		public int maxPoints(Point[] p) {
			if (p == null || p.length < 2) return p == null ? 0 : p.length;
			HashMap<Pair, Integer> map = new HashMap<Pair, Integer>();
			int ans = 0;
			for (int i = 0; i < p.length; i ++) {
				map.clear();
				int same = 0;
				int max = 1;
				for (int j = 0; j < p.length; j ++) {
					if (i == j) continue;
					if (p[i].x == p[j].x && p[i].y == p[j].y) {
						same ++;
						continue;
					}
					Pair pair = new Pair(p[j].x - p[i].x, p[j].y - p[i].y);
					Integer v = map.get(pair);
					int nv = v == null ? 2 : v + 1;
					max = Math.max(max, nv);
					map.put(pair, nv);
				}
				ans = Math.max(ans, max + same);
			}
			return ans;
		}
		static class Pair {
		    int a, b;
		    public Pair(int a, int b) {
		        int c = gcd(Math.abs(a), Math.abs(b));
		        a = a / c;
		        b = b / c;
		        if (a < 0 && b < 0) {
		            a = -a;
		            b = -b;
		        } else if (((a >>> 31 ) ^ (b >>> 31)) == 1) {
		            a = -Math.abs(a);
		            b = Math.abs(b);
		        }
		        this.a = a;
		        this.b = b;
            }
		    private int gcd(int a, int b) {
		        if (a < b) return gcd(b, a);
		        int t = 0;
		        while (b != 0) {
		            t = a % b;
		            a = b;
		            b = t;
		        }
		        return a;
		    }
            @Override
            public boolean equals(Object obj) {
                if (! (obj instanceof Pair)) return false;
                Pair p = (Pair) obj;
                return this.a == p.a && this.b == p.b;
            }
            @Override
            public int hashCode() {
                return this.a * 31 + this.b;
            }
		}
	}
	static class Point {
		int x;
		int y;
		Point() { x = 0; y = 0; }
		Point(int a, int b) { x = a; y = b; }
	}
    static class Solution2 {
        public int maxPoints(Point[] p) {
            if (p == null || p.length < 2) return p == null ? 0 : p.length;
            HashMap<Double, Integer> map = new HashMap<Double, Integer>();
            int ans = 0;
            for (int i = 0; i < p.length; i ++) {
                map.clear();
                int same = 0;
                int max = 1;
                for (int j = 0; j < p.length; j ++) {
                    if (i == j) continue;
                    if (p[i].x == p[j].x && p[i].y == p[j].y) {
                        same ++;
                        continue;
                    }
                    double pair = calc(p[j].x - p[i].x, p[j].y - p[i].y);
                    Integer v = map.get(pair);
                    int nv = v == null ? 2 : v + 1;
                    max = Math.max(max, nv);
                    map.put(pair, nv);
                }
                ans = Math.max(ans, max + same);
            }
            return ans;
        }
        private double calc(int a, int b) {
            int c = gcd(Math.abs(a), Math.abs(b));
            a = a / c;
            b = b / c;
            if (a < 0 && b < 0) {
                a = -a;
                b = -b;
            } else if (((a >>> 31 ) ^ (b >>> 31)) == 1) {
                a = -Math.abs(a);
                b = Math.abs(b);
            }
            double ans = 0.0;
            ans += a;
            ans = ans * 200000000;
            ans += b;
            return ans;
        }
        private int gcd(int a, int b) {
            if (a < b) return gcd(b, a);
            int t = 0;
            while (b != 0) {
                t = a % b;
                a = b;
                b = t;
            }
            return a;
        }
    }
}

File: python/leetcode/P149_MaxPointsOnALine.py
"""
User:  zxwtry
File:  P149_MaxPointsOnALine.py
Date:  2018/4/2
Time:  21:59
"""

from util.Point import Point

class Solution:
    def maxPoints(self, points):
        """
        :type points: List[Point]
        :rtype: int
        """
        a1, a2, a3 = list(), list(), list()
        # 查看重复的点
        p1, p2 = None, None
        ps = list()
        ps_cnt = list()
        for p in points:
            exists = False
            for ps_i in range(len(ps)):
                if self.judgeEqual(p, ps[ps_i]):
                    ps_cnt[ps_i] += 1
                    exists = True
            if not exists:
                ps.append(p)
                ps_cnt.append(1)
        if len(ps) < 2:
            return sum(cnt for cnt in ps_cnt)
        ans, ps_len = 2, len(ps)
        for i in range(ps_len):
            for j in range(i + 1, ps_len, 1):
                ans_one = ps_cnt[i] + ps_cnt[j]
                for k in range(ps_len):
                    if k == i or k == j:
                        continue
                    if self.judgeOnline(ps[i], ps[j], ps[k]):
                        ans_one += ps_cnt[k]
                ans = max(ans, ans_one)
        return ans

    def judgeOnline(self, p1, p2, p3):
        return (p2.y - p1.y) * (p3.x - p1.x) == (p2.x - p1.x) * (p3.y - p1.y)

    def judgeEqual(self, p1, p2):
        if p1 is None and p2 is None:
            return True
        if p1 is None or p2 is None:
            return False
        return p1.x == p2.x and p1.y == p2.y


class Solution2:
    def maxPoints(self, points):
        """
        :type points: List[Point]
        :rtype: int
        """
        a1, a2, a3 = list(), list(), list()
        # 查看重复的点
        p1, p2 = None, None
        ps = list()
        ps_cnt = list()
        for p in points:
            exists = False
            for ps_i in range(len(ps)):
                if self.judgeEqual(p, ps[ps_i]):
                    ps_cnt[ps_i] += 1
                    exists = True
            if not exists:
                ps.append(p)
                ps_cnt.append(1)
        if len(ps) < 2:
            return sum(cnt for cnt in ps_cnt)
        ans, ps_len = 2, len(ps)
        for i in range(ps_len):
            m = dict()
            for j in range(ps_len):
                if i == j:
                    continue
                xv = ps[i].x - ps[j].x
                yv = ps[i].y - ps[j].y
                gcdv = self.gcd(xv, yv)
                xv = xv // gcdv
                yv = yv // gcdv
                if xv < 0:
                    xv, yv = -xv, -yv
                key = str(xv) + "#" + str(yv)
                if key in m.keys():
                    m[key] += ps_cnt[j]
                else:
                    m[key] = ps_cnt[j]
            ans = max(ans, max(v for v in m.values()) + ps_cnt[i])
        return ans

    def judgeEqual(self, p1, p2):
        if p1 is None and p2 is None:
            return True
        if p1 is None or p2 is None:
            return False
        return p1.x == p2.x and p1.y == p2.y

    def gcd(self, a, b):
        a, b = abs(a), abs(b)
        a, b = max(a, b), min(a, b)
        while b != 0:
            a, b = b, a % b
        return a

if __name__ == "__main__":
    # [[0,0],[1,65536],[65536,0]]
    xs = [0, 1, 65536]
    ys = [0, 65536, 0]
    ps = Point().constructFromArray(xs, ys)
    print(Solution2().maxPoints(ps))
